Failed to produce diagnostic: resultBuilder containing switch containing invalid case with an associated value

### Description

Per the title, see example. When a resultBuilder contains an `if case … = …` or a `switch` where the case is both invalid and has an associated object, the compiler displays the "Failed to produce diagnostic for expression; please submit a bug report" message. Removing the resultBuilder, correcting the case, or removing the associated value (whether or not empty parentheses remain) each fix the issue and behave as expected.

I assume it applies to all resultBuilders, though I've only actually tested `@ViewBuilder`.

### Reproduction

`if case`:
```swift
import SwiftUI
enum E {
    case e
}
@ViewBuilder
func foo(_ e: E) { // error: failed to produce diagnostic for expression; please submit a bug report
    if case .a(.x) = e {
        EmptyView()
    }
}
```

`switch`:
```swift
import SwiftUI
enum E {
    case e
}
@ViewBuilder
func foo(_ e: E) { // error: failed to produce diagnostic for expression; please submit a bug report
    switch e {
    case .a(.x):
        EmptyView()
    }
}
```

Complete command line example:
```zsh
----------% cat -n test.swift
     1	import SwiftUI
     2	enum E {
     3	    case e
     4	}
     5	@ViewBuilder
     6	func foo(_ e: E) {
     7	    if case .a(.x) = e {
     8	        EmptyView()
     9	    }
    10	}
----------% xcrun swiftc --version
swift-driver version: 1.127.5.3 Apple Swift version 6.2 (swiftlang-6.2.0.10.950 clang-1700.3.10.950)
Target: arm64-apple-macosx15.0
----------% xcrun -l swiftc test.swift
env SDKROOT=/Applications/Xcode-26b2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.0.sdk /Applications/Xcode-26b2.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc test.swift
test.swift:6:6: error: failed to produce diagnostic for expression; please submit a bug report (https://swift.org/contributing/#reporting-bugs)
 4 | }
 5 | @ViewBuilder
 6 | func foo(_ e: E) {
   |      `- error: failed to produce diagnostic for expression; please submit a bug report (https://swift.org/contributing/#reporting-bugs)
 7 |     if case .a(.x) = e {
 8 |         EmptyView()
----------%                                                                                                                                                1
```

### Expected behavior

Compiler should print a useful diagnostic, likely the same as outside the resultBuilder context.
For `if case…`:
- Type of expression is ambiguous without a type annotation
For `switch`:
- Cannot infer contextual base in reference to member 'x'
- Type 'E' has no member 'a'

### Environment

swift-driver version: 1.127.5.3 Apple Swift version 6.2 (swiftlang-6.2.0.10.950 clang-1700.3.10.950)
Target: arm64-apple-macosx15.0

Occurs at least as far back as I have Xcode installed:
swift-driver version: 1.115.1 Apple Swift version 6.0.3 (swiftlang-6.0.3.1.10 clang-1600.0.30.1)
Target: arm64-apple-macosx15.0

### Additional information

_No response_

--- a/include/swift/AST/DiagnosticsSema.def
+++ b/include/swift/AST/DiagnosticsSema.def
@@ -1801,6 +1801,13 @@ ERROR(too_many_dynamic_member_lookups,none,
       "dynamic member lookups",
       (DeclNameRef))
 
+ERROR(keypath_dynamic_member_lookup_expects_class_base, none,
+      "referencing %kind1 on %0 requires it to be a class type",
+      (Type, ValueDecl *))
+NOTE(keypath_dynamic_member_lookup_expects_class_base_note, none,
+     "'subscript(dynamicMember:)' declared in %0 uses 'ReferenceWritableKeyPath'"
+     " which only supports classes", (Type))
+
 ERROR(string_index_not_integer,none,
       "String must not be indexed with %0, it has variable size elements",
       (Type))

--- a/include/swift/Sema/CSFix.h
+++ b/include/swift/Sema/CSFix.h
@@ -500,6 +500,10 @@ enum class FixKind : uint8_t {
 
   /// Ignore that a conformance is isolated but is not allowed to be.
   IgnoreIsolatedConformance,
+
+  /// Ignore passing `WritableKeyPath` to `ReferenceWritableKeyPath` mismatch
+  /// when trying to access a member using key path dynamic member lookup.
+  IgnoreClassRequirementForDynamicMemberLookup,
 };
 
 enum class FixImpact : unsigned {
@@ -4032,6 +4036,34 @@ public:
 
   static bool classof(const ConstraintFix *fix) {
     return fix->getKind() == FixKind::IgnoreIsolatedConformance;
+  }
+};
+
+class IgnoreClassRequirementForDynamicMemberLookup : public ConstraintFix {
+  Type BaseType;
+  ValueDecl *Member;
+
+  IgnoreClassRequirementForDynamicMemberLookup(ConstraintSystem &cs,
+                                               Type baseTy, ValueDecl *member,
+                                               ConstraintLocator *locator)
+      : ConstraintFix(cs, FixKind::IgnoreClassRequirementForDynamicMemberLookup,
+                      locator),
+        BaseType(baseTy), Member(member) {}
+
+public:
+  std::string getName() const override {
+    return "ignore non-class base used for key path dynamic member lookup";
+  }
+
+  bool diagnose(const Solution &solution, bool asNote = false) const override;
+
+  static IgnoreClassRequirementForDynamicMemberLookup *
+  create(ConstraintSystem &cs, Type baseTy, ValueDecl *member,
+         ConstraintLocator *locator);
+
+  static bool classof(const ConstraintFix *fix) {
+    return fix->getKind() ==
+           FixKind::IgnoreClassRequirementForDynamicMemberLookup;
   }
 };
 

--- a/include/swift/Sema/TypeVariableType.h
+++ b/include/swift/Sema/TypeVariableType.h
@@ -228,6 +228,10 @@ public:
   /// operator.
   bool isTernary() const;
 
+  /// Determine whether this type variable represents a type of a synthesized
+  /// argument of a call/subscript.
+  bool isSynthesizedArgument() const;
+
   /// Retrieve the representative of the equivalence class to which this
   /// type variable belongs.
   ///

--- a/lib/Sema/BindingProducer.cpp
+++ b/lib/Sema/BindingProducer.cpp
@@ -570,13 +570,14 @@ TypeVariableBinding::fixForHole(ConstraintSystem &cs) const {
         // completion token at all.
         return std::nullopt;
       }
+
       // Not being able to infer the type of a variable in a pattern binding
       // decl is more dramatic than anything that could happen inside the
       // expression because we want to preferrably point the diagnostic to a
       // part of the expression that caused us to be unable to infer the
       // variable's type.
       ConstraintFix *fix =
-          IgnoreUnresolvedPatternVar::create(cs, pattern.get(), dstLocator);
+        IgnoreUnresolvedPatternVar::create(cs, pattern.get(), dstLocator);
       return std::make_pair(fix, FixImpact::InvalidAST * 10);
     }
   }

--- a/lib/Sema/BindingProducer.cpp
+++ b/lib/Sema/BindingProducer.cpp
@@ -660,30 +660,6 @@ bool TypeVariableBinding::attempt(ConstraintSystem &cs) const {
   if (type->hasJoinOrMeet())
     type = openTypeJoinsAndMeets(cs, type, dstLocator);
 
-  // If type variable has been marked as a possible hole due to
-  // e.g. reference to a missing member. Let's propagate that
-  // information to the object type of the optional type it's
-  // about to be bound to.
-  //
-  // In some situations like pattern bindings e.g. `if let x = base?.member`
-  // - if `member` doesn't exist, `x` cannot be determined either, which
-  // leaves `OptionalEvaluationExpr` representing outer type of `base?.member`
-  // without any contextual information, so even though `x` would get
-  // bound to result type of the chain, underlying type variable wouldn't
-  // be resolved, so we need to propagate holes up the conversion chain.
-  // Also propagate in code completion mode because in some cases code
-  // completion relies on type variable being a potential hole.
-  if (TypeVar->getImpl().canBindToHole()) {
-    if (srcLocator->directlyAt<OptionalEvaluationExpr>() ||
-        cs.isForCodeCompletion()) {
-      if (auto objectTy = type->getOptionalObjectType()) {
-        if (auto *typeVar = objectTy->getAs<TypeVariableType>()) {
-          cs.recordPotentialHole(typeVar);
-        }
-      }
-    }
-  }
-
   ConstraintSystem::TypeMatchOptions options;
 
   options |= ConstraintSystem::TMF_GenerateConstraints;

--- a/lib/Sema/CSDiagnostics.cpp
+++ b/lib/Sema/CSDiagnostics.cpp
@@ -9931,3 +9931,19 @@ bool DisallowedIsolatedConformance::diagnoseAsError() {
 
   return true;
 }
+
+bool NonClassBaseInDynamicMemberLookup::diagnoseAsError() {
+  emitDiagnostic(diag::keypath_dynamic_member_lookup_expects_class_base,
+                 BaseType, Member);
+
+  auto *loc = getLocator();
+  auto *memberLoc =
+      getConstraintLocator(loc->getAnchor(), loc->getPath().drop_back());
+
+  if (auto overload = getCalleeOverloadChoiceIfAvailable(memberLoc)) {
+    emitDiagnostic(diag::keypath_dynamic_member_lookup_expects_class_base_note,
+                   overload->choice.getBaseType());
+  }
+
+  return true;
+}

--- a/lib/Sema/CSDiagnostics.h
+++ b/lib/Sema/CSDiagnostics.h
@@ -3418,6 +3418,23 @@ public:
   bool diagnoseAsError() override;
 };
 
+/// Diagnose situations when a key path dynamic member lookup expects
+/// a `ReferenceWritableKeyPath` but the base is not a class and so
+/// the argument is `WritableKeyPath`.
+class NonClassBaseInDynamicMemberLookup final : public FailureDiagnostic {
+  Type BaseType;
+  ValueDecl *Member;
+
+public:
+  NonClassBaseInDynamicMemberLookup(const Solution &solution, Type baseType,
+                                    ValueDecl *member,
+                                    ConstraintLocator *locator)
+      : FailureDiagnostic(solution, locator), BaseType(resolveType(baseType)),
+        Member(member) {}
+
+  bool diagnoseAsError() override;
+};
+
 } // end namespace constraints
 } // end namespace swift
 

--- a/lib/Sema/CSFix.cpp
+++ b/lib/Sema/CSFix.cpp
@@ -2900,3 +2900,18 @@ bool IgnoreIsolatedConformance::diagnose(const Solution &solution,
   DisallowedIsolatedConformance failure(solution, conformance, getLocator());
   return failure.diagnose(asNote);
 }
+
+IgnoreClassRequirementForDynamicMemberLookup *
+IgnoreClassRequirementForDynamicMemberLookup::create(
+    ConstraintSystem &cs, Type baseTy, ValueDecl *member,
+    ConstraintLocator *locator) {
+  return new (cs.getAllocator())
+      IgnoreClassRequirementForDynamicMemberLookup(cs, baseTy, member, locator);
+}
+
+bool IgnoreClassRequirementForDynamicMemberLookup::diagnose(
+    const Solution &solution, bool asNote) const {
+  NonClassBaseInDynamicMemberLookup failure(solution, BaseType, Member,
+                                            getLocator());
+  return failure.diagnose(asNote);
+}

--- a/lib/Sema/CSGen.cpp
+++ b/lib/Sema/CSGen.cpp
@@ -4238,6 +4238,13 @@ bool ConstraintSystem::generateConstraints(
           target.getPatternBindingOfUninitializedVar(),
           target.getIndexOfUninitializedVar());
 
+      // If `typeCheckPattern` produced a type with errors, let's turn pattern
+      // type into a hole to avoid a recording fallback fix.
+      if (target.getTypeOfUninitializedVar()->hasError()) {
+        increaseScore(SK_Hole, locator);
+        recordTypeVariablesAsHoles(patternType);
+      }
+
       return !patternType;
     }
   }

--- a/lib/Sema/CSGen.cpp
+++ b/lib/Sema/CSGen.cpp
@@ -2435,9 +2435,14 @@ namespace {
         auto enumPattern = cast<EnumElementPattern>(pattern);
 
         // Create a type variable to represent the pattern.
+        //
+        // This type is contextually dependent in cases like `.a` where
+        // `patternType` represents an implicit base type of `.a` and so
+        // should be allowed to bind to a hole just like regular leading-dot
+        // syntax.
         Type patternType =
             CS.createTypeVariable(CS.getConstraintLocator(locator),
-                                  TVO_CanBindToNoEscape);
+                                  TVO_CanBindToNoEscape | TVO_CanBindToHole);
 
         // Form the member constraint for a reference to a member of this
         // type.

--- a/lib/Sema/CSGen.cpp
+++ b/lib/Sema/CSGen.cpp
@@ -3124,9 +3124,16 @@ namespace {
       // like this to be the smallest possible nesting level of
       // optional types, e.g. T? over T??; otherwise we don't really
       // have a preference.
-      auto valueTy = CS.createTypeVariable(CS.getConstraintLocator(expr),
-                                           TVO_PrefersSubtypeBinding |
-                                           TVO_CanBindToNoEscape);
+      unsigned options = TVO_PrefersSubtypeBinding | TVO_CanBindToNoEscape;
+
+      // In completion mode we don't care about fixes, let's allow
+      // the result of an optional chain to be bound to a hole to
+      // for a solution for situations like `a?<#token#>.b`.
+      if (CS.isForCodeCompletion())
+        options |= TVO_CanBindToHole;
+
+      auto valueTy =
+          CS.createTypeVariable(CS.getConstraintLocator(expr), options);
 
       Type optTy = getOptionalType(expr->getSubExpr()->getLoc(), valueTy);
       if (!optTy)

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -4342,7 +4342,8 @@ ConstraintSystem::matchExistentialTypes(Type type1, Type type2,
         // If the path ends at `optional payload` it means that this
         // check is part of an implicit value-to-optional conversion,
         // and it could be safely dropped.
-        if (!path.empty() && path.back().is<LocatorPathElt::OptionalInjection>())
+        while (!path.empty() &&
+               path.back().is<LocatorPathElt::OptionalInjection>())
           path.pop_back();
 
         // Determine whether this conformance mismatch is
@@ -4404,27 +4405,44 @@ ConstraintSystem::matchExistentialTypes(Type type1, Type type2,
           if (last.is<LocatorPathElt::InstanceType>())
             return SolutionKind::Error;
 
+          // Fail here and let the generic argument matching handle the failure
+          // because we need to present complete types in the diagnostic and
+          // point to the mismatching location(s).
+          if (last.is<LocatorPathElt::GenericArgument>())
+            return SolutionKind::Error;
+
+          if (last.is<LocatorPathElt::CoercionOperand>()) {
+            auto *fix = ContextualMismatch::create(
+                *this, type1, type2, getConstraintLocator(anchor, path));
+            if (recordFix(fix))
+              return SolutionKind::Error;
+            break;
+          }
+
+          if (last.is<LocatorPathElt::ContextualType>() ||
+              last.is<LocatorPathElt::EnumPatternImplicitCastMatch>() ||
+              last.is<LocatorPathElt::SequenceElementType>()) {
+            auto proto = protoDecl->getDeclaredInterfaceType();
+            auto *fix = MissingConformance::forContextual(
+                *this, type1, proto, getConstraintLocator(anchor, path));
+
+            if (recordFix(fix))
+              return SolutionKind::Error;
+
+            break;
+          }
         } else { // There are no elements in the path
-          if (!(isExpr<AssignExpr>(anchor) || isExpr<CoerceExpr>(anchor)))
-            return SolutionKind::Error;
-        }
-
-        if (isExpr<CoerceExpr>(anchor)) {
-          auto *fix = ContextualMismatch::create(
-              *this, type1, type2, getConstraintLocator(anchor, path));
-          if (recordFix(fix))
-            return SolutionKind::Error;
-          break;
-        }
-
-        auto proto = protoDecl->getDeclaredInterfaceType();
-        auto *fix = MissingConformance::forContextual(
-            *this, type1, proto, getConstraintLocator(anchor, path));
-
-        if (recordFix(fix))
+          if (isExpr<AssignExpr>(anchor)) {
+            auto *fix = ContextualMismatch::create(
+                *this, type1, type2, getConstraintLocator(anchor, path));
+            if (recordFix(fix))
+              return SolutionKind::Error;
+            break;
+          }
           return SolutionKind::Error;
+        }
 
-        break;
+        return SolutionKind::Error;
       }
     }
   }

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -40,6 +40,7 @@
 #include "swift/ClangImporter/ClangModule.h"
 #include "swift/Sema/CSFix.h"
 #include "swift/Sema/Constraint.h"
+#include "swift/Sema/ConstraintLocator.h"
 #include "swift/Sema/ConstraintSystem.h"
 #include "swift/Sema/IDETypeChecking.h"
 #include "swift/Sema/PreparedOverload.h"
@@ -6846,6 +6847,9 @@ bool ConstraintSystem::repairFailures(
     if (lhs->isPlaceholder() || rhs->isPlaceholder())
       return true;
 
+    if (hasAnyRestriction())
+      return false;
+
     if (repairViaOptionalUnwrap(*this, lhs, rhs, matchKind, conversionsOrFixes,
                                 locator))
       return true;
@@ -6856,6 +6860,12 @@ bool ConstraintSystem::repairFailures(
         conversionsOrFixes.push_back(
             CollectionElementContextualMismatch::create(
                 *this, lhs, rhs, getConstraintLocator(anchor, path)));
+        return true;
+      }
+      if (path.back().is<LocatorPathElt::ApplyArgToParam>()) {
+        conversionsOrFixes.push_back(
+            AllowArgumentMismatch::create(*this, lhs, rhs,
+                                          getConstraintLocator(anchor, path)));
         return true;
       }
     }
@@ -14646,6 +14656,27 @@ ConstraintSystem::simplifyRestrictedConstraintImpl(
           llvm::PointerIntPair<Type, 3, unsigned> baseType2) -> SolutionKind {
     if (restriction != ConversionRestrictionKind::PointerToPointer)
       increaseScore(ScoreKind::SK_ValueToPointerConversion, locator);
+
+    // Since this match doesn't add any custom path elements, let's drop
+    // `OptionalInjection` to avoid confusing generic argument matching.
+    // This is important, for example, in situations like
+    // `inout T` -> `UnsafePointer<U>?` in argument positions,
+    // `inout T` would be "injected" into optional by unwrapping `UnsafePointer`
+    // so the only match that matters here is pointers and their
+    // `Pointee` generic parameters.
+    if (shouldAttemptFixes()) {
+      if (locator.endsWith<LocatorPathElt::OptionalInjection>()) {
+        SmallVector<LocatorPathElt, 2> path;
+        auto anchor = locator.getLocatorParts(path);
+
+        while (!path.empty() &&
+               path.back().is<LocatorPathElt::OptionalInjection>()) {
+          path.pop_back();
+        }
+
+        locator = getConstraintLocator(anchor, path);
+      }
+    }
 
     auto result =
         matchTypes(baseType1.getPointer(), baseType2.getPointer(),

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -11539,10 +11539,6 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
   // If the base type of this member lookup is a "hole" there is no
   // reason to perform a lookup because it wouldn't return any results.
   if (shouldAttemptFixes()) {
-    auto markMemberTypeAsPotentialHole = [&](Type memberTy) {
-      recordAnyTypeVarAsPotentialHole(simplifyType(memberTy));
-    };
-
     // If this is an unresolved member ref e.g. `.foo` and its contextual base
     // type has been determined to be a "hole", let's mark the resulting member
     // type as a potential hole and continue solving.
@@ -11594,14 +11590,14 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
             return SolutionKind::Error;
         }
 
-        markMemberTypeAsPotentialHole(memberTy);
+        recordTypeVariablesAsHoles(memberTy);
         return SolutionKind::Solved;
       }
     } else if (kind == ConstraintKind::ValueMember &&
                baseObjTy->getMetatypeInstanceType()->isPlaceholder()) {
       // If base type is a "hole" there is no reason to record any
       // more "member not found" fixes for chained member references.
-      markMemberTypeAsPotentialHole(memberTy);
+      recordTypeVariablesAsHoles(memberTy);
       return SolutionKind::Solved;
     }
   }

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -39,6 +39,7 @@
 #include "swift/Basic/StringExtras.h"
 #include "swift/ClangImporter/ClangModule.h"
 #include "swift/Sema/CSFix.h"
+#include "swift/Sema/Constraint.h"
 #include "swift/Sema/ConstraintSystem.h"
 #include "swift/Sema/IDETypeChecking.h"
 #include "swift/Sema/PreparedOverload.h"
@@ -6896,6 +6897,11 @@ bool ConstraintSystem::repairFailures(
   }
 
   case ConstraintLocator::PatternMatch: {
+    // Let's let the matching (i.e. deep equality) happen before attempting any
+    // fixes for patterns.
+    if (hasAnyRestriction())
+      return false;
+
     auto *pattern = elt.castTo<LocatorPathElt::PatternMatch>().getPattern();
 
     // TODO: We ought to introduce a new locator element for this.

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -4820,6 +4820,19 @@ ConstraintSystem::matchTypesBindTypeVar(
                : SolutionKind::Error;
   }
 
+  // If a type assigned to a synthesized argument has type variables
+  // it should be possible to assign holes to them because it's not
+  // guaranteed that they could always be resolved from the context.
+  //
+  // For example, when a parameter has a function type that returns
+  // a generic parameter, if it's not connected to anything else i.e.
+  // doesn't appear in other parameter positions or there is no
+  // contextual type for the call when it's a result, it won't be
+  // possible to resolve.
+  if (typeVar->getImpl().isSynthesizedArgument()) {
+    recordAnyTypeVarAsPotentialHole(type);
+  }
+
   assignFixedType(typeVar, type, /*updateState=*/true,
                   /*notifyInference=*/!flags.contains(TMF_BindingTypeVariable));
 

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -6798,6 +6798,9 @@ bool ConstraintSystem::repairFailures(
     if (lhs->isPlaceholder() || rhs->isPlaceholder())
       return true;
 
+    if (hasAnyRestriction())
+      return false;
+
     if (repairViaOptionalUnwrap(*this, lhs, rhs, matchKind, conversionsOrFixes,
                                 locator))
       return true;
@@ -14696,6 +14699,7 @@ ConstraintSystem::simplifyRestrictedConstraintImpl(
     if (loc->isForAssignment() || loc->isForCoercion() ||
         loc->isForContextualType() ||
         loc->isLastElement<LocatorPathElt::ApplyArgToParam>() ||
+        loc->isLastElement<LocatorPathElt::UnresolvedMemberChainResult>() ||
         loc->isForOptionalTry()) {
       if (restriction == ConversionRestrictionKind::Superclass) {
         if (auto *fix = CoerceToCheckedCast::attempt(

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -11424,7 +11424,8 @@ static ConstraintFix *fixMemberRef(
     case MemberLookupResult::UR_WritableKeyPathOnReadOnlyMember:
       return TreatRValueAsLValue::create(cs, cs.getConstraintLocator(locator));
     case MemberLookupResult::UR_ReferenceWritableKeyPathOnMutatingMember:
-      break;
+      return IgnoreClassRequirementForDynamicMemberLookup::create(
+          cs, baseTy, choice.getDecl(), cs.getConstraintLocator(locator));
     case MemberLookupResult::UR_KeyPathWithAnyObjectRootType:
       return AllowAnyObjectKeyPathRoot::create(cs, locator);
 
@@ -16123,6 +16124,7 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
   case FixKind::TooManyDynamicMemberLookups:
   case FixKind::IgnoreNonMetatypeDynamicType:
   case FixKind::IgnoreIsolatedConformance:
+  case FixKind::IgnoreClassRequirementForDynamicMemberLookup:
     llvm_unreachable("handled elsewhere");
   }
 

--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -5073,7 +5073,8 @@ repairViaOptionalUnwrap(ConstraintSystem &cs, Type fromType, Type toType,
   // `let _: Bool = try? foo()` and `foo()` produces `Int`
   // we should diagnose it as type mismatch instead of missing unwrap.
   bool possibleContextualMismatch = [&]() {
-    if (!locator.endsWith<LocatorPathElt::ContextualType>())
+    if (!(locator.endsWith<LocatorPathElt::ContextualType>() ||
+          locator.endsWith<LocatorPathElt::Condition>()))
       return false;
 
     // If the contextual type is optional as well, it's definitely a

--- a/lib/Sema/CSSyntacticElement.cpp
+++ b/lib/Sema/CSSyntacticElement.cpp
@@ -636,6 +636,19 @@ private:
         ContextualPattern::forPatternBindingDecl(patternBinding, index);
     Type patternType = TypeChecker::typeCheckPattern(contextualPattern);
 
+    if (patternType->hasError()) {
+      // Fail immediately if this is an invalid pattern and the solver is not
+      // in the diagnostic mode.
+      if (!cs.shouldAttemptFixes()) {
+        hadError = true;
+        return;
+      }
+
+      // In diagnostic mode, let's let the type through. It would get pattern
+      // type turned into a hole during the constraint generation for the
+      // pattern.
+    }
+
     auto target = getTargetForPattern(patternBinding, index, patternType);
     if (!target) {
       hadError = true;

--- a/lib/Sema/TypeVariableType.cpp
+++ b/lib/Sema/TypeVariableType.cpp
@@ -163,6 +163,11 @@ bool TypeVariableType::Implementation::isTernary() const {
   return locator && locator->directlyAt<TernaryExpr>();
 }
 
+bool TypeVariableType::Implementation::isSynthesizedArgument() const {
+  return locator &&
+         locator->isLastElement<LocatorPathElt::SynthesizedArgument>();
+}
+
 TypeVariableType *ConstraintSystem::createTypeVariable(
                                      ConstraintLocator *locator,
                                      unsigned options,

--- a/test/Constraints/binding_order.swift
+++ b/test/Constraints/binding_order.swift
@@ -101,9 +101,10 @@ do {
     // Some of these might be hard to resolve, but we should produce better diagnostics.
 
     let _: [any Command] = [aOpt, bOpt].filter { $0 != nil }
-    // expected-error@-1 {{no exact matches in call to instance method 'filter'}}
+    // expected-error@-1 {{cannot convert value of type 'A?' to expected element type 'any Command'}}
+    // expected-error@-2 {{cannot convert value of type 'B?' to expected element type 'any Command'}}
     let _: [any Command] = [a, b].map { $0 }
-    // expected-error@-1 {{failed to produce diagnostic for expression}}
+    // expected-error@-1 {{cannot convert value of type 'Super' to closure result type 'any Command'}}
     let _: [any Command] = [a, b].flatMap { [$0] }
     // expected-error@-1 {{cannot convert value of type 'Super' to expected element type 'any Command'}}
 

--- a/test/Constraints/closures.swift
+++ b/test/Constraints/closures.swift
@@ -1175,8 +1175,7 @@ func rdar76058892() {
   func experiment(arr: [S]?) {
     test { // expected-error {{contextual closure type '() -> String' expects 0 arguments, but 1 was used in closure body}}
       if let arr = arr {
-        arr.map($0.test) // expected-note {{anonymous closure parameter '$0' is used here}} // expected-error {{generic parameter 'T' could not be inferred}}
-        // expected-error@-1 {{generic parameter 'E' could not be inferred}}
+        arr.map($0.test) // expected-note {{anonymous closure parameter '$0' is used here}}
       }
     }
   }

--- a/test/Constraints/diag_missing_arg.swift
+++ b/test/Constraints/diag_missing_arg.swift
@@ -193,3 +193,13 @@ do {
     // expected-error@-1 {{cannot convert value of type '() -> Void' to expected argument type '(Int) -> Void'}}
   }
 }
+
+// https://github.com/swiftlang/swift/issues/83911
+do {
+  func compute<R>(_: Int = 42, _: () -> R) -> R {} // expected-note {{'compute' declared here}}
+  func compute<R>(_: Int = 42, _: () -> R, completion: @escaping () -> Void) -> R {}
+
+  func test() {
+    compute() // expected-error {{missing argument for parameter #2 in call}}
+  }
+}

--- a/test/Constraints/fixes.swift
+++ b/test/Constraints/fixes.swift
@@ -361,7 +361,4 @@ func testKeyPathSubscriptArgFixes(_ fn: @escaping () -> Int) {
 // https://github.com/apple/swift/issues/54865
 func f_54865(a: Any, _ str: String?) {
   a == str  // expected-error {{cannot convert value of type 'Any' to expected argument type 'String'}}
-  // expected-error@-1 {{value of optional type 'String?' must be unwrapped to a value of type 'String}}
-  // expected-note@-2 {{coalesce using '??' to provide a default when the optional value contains 'nil'}}
-  // expected-note@-3 {{force-unwrap using '!' to abort execution if the optional value contains 'nil'}}
 }

--- a/test/Constraints/keypath_dynamic_member_lookup_errors.swift
+++ b/test/Constraints/keypath_dynamic_member_lookup_errors.swift
@@ -1,0 +1,26 @@
+// RUN: %target-typecheck-verify-swift
+
+@dynamicMemberLookup
+struct ObservedObject<T> {
+  let root: T
+
+  subscript<U>(
+    dynamicMember keyPath: ReferenceWritableKeyPath<T, U>
+  ) -> U {
+    fatalError()
+  }
+}
+
+protocol P {
+  var checked: Bool { get set }
+}
+
+struct Test<U: P> {
+  var v: ObservedObject<U>
+
+  func test() {
+    v.checked
+    // expected-error@-1 {{referencing property 'checked' on 'U' requires it to be a class type}}
+    // expected-note@-2 {{'subscript(dynamicMember:)' declared in 'ObservedObject<U>' uses 'ReferenceWritableKeyPath' which only supports classes}}
+  }
+}

--- a/test/Constraints/optional.swift
+++ b/test/Constraints/optional.swift
@@ -652,3 +652,14 @@ do {
     context(from: TestSyntax(v) ?? TestSyntax(other.first?.value), root: v) // Ok (no warnings)
   }
 }
+
+do {
+  struct Test<T> { // expected-note {{'T' declared as parameter to type 'Test'}}
+    var v: T
+    init?(_: Int) {
+    }
+  }
+  _ = Test(0)?.v
+  // expected-error@-1 {{generic parameter 'T' could not be inferred}}
+  // expected-note@-2 {{explicitly specify the generic arguments to fix this issue}}
+}

--- a/test/Constraints/patterns.swift
+++ b/test/Constraints/patterns.swift
@@ -714,38 +714,27 @@ func testLotsOfNil(_ x: LotsOfOptional) {
 func issue66752(_ x: Result<String, Error>) {
   let _ = {
     if case .failure() = x {}
-    // expected-error@-1 {{type '()' cannot conform to 'Error}}
-    // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
-    // expected-note@-3 {{required by generic enum 'Result' where 'Failure' = '()'}}
+    // expected-error@-1 {{missing pattern #1 in enum associated value match}}
   }
   let _ = {
     if case (.failure(), let y) = (x, 0) {}
-    // expected-error@-1 {{type '()' cannot conform to 'Error}}
-    // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
-    // expected-note@-3 {{required by generic enum 'Result' where 'Failure' = '()'}}
+    // expected-error@-1 {{missing pattern #1 in enum associated value match}}
   }
   if case .failure() = x {}
-  // expected-error@-1 {{type '()' cannot conform to 'Error}}
-  // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
-  // expected-note@-3 {{required by generic enum 'Result' where 'Failure' = '()'}}
+  // expected-error@-1 {{missing pattern #1 in enum associated value match}}
   
   if case (.failure(), let y) = (x, 0) {}
-  // expected-error@-1 {{type '()' cannot conform to 'Error}}
-  // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
-  // expected-note@-3 {{required by generic enum 'Result' where 'Failure' = '()'}}
+  // expected-error@-1 {{missing pattern #1 in enum associated value match}}
 }
 
 // https://github.com/apple/swift/issues/66750
-// FIXME: We ought to improve the diagnostics here
 func issue66750(_ x: Result<String, Error>) {
   let _ = {
     switch x {
     case .success:
       "a"
     case .failure():
-      // expected-error@-1 {{type '()' cannot conform to 'Error}}
-      // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
-      // expected-note@-3 {{required by generic enum 'Result' where 'Failure' = '()'}}
+      // expected-error@-1 {{missing pattern #1 in enum associated value match}}
       "b"
     }
   }
@@ -754,24 +743,35 @@ func issue66750(_ x: Result<String, Error>) {
     case (.success, let y):
       "a"
     case (.failure(), let y):
-      // expected-error@-1 {{type '()' cannot conform to 'Error}}
-      // expected-note@-2 {{only concrete types such as structs, enums and classes can conform to protocols}}
-      // expected-note@-3 {{required by generic enum 'Result' where 'Failure' = '()'}}
+      // expected-error@-1 {{missing pattern #1 in enum associated value match}}
       "b"
     }
   }
   switch x {
   case .success:
     break
-  case .failure(): // expected-error {{tuple pattern cannot match values of the non-tuple type 'any Error'}}
+  // FIXME: We ought to improve the diagnostics here
+  case .failure():
+    // expected-error@-1 {{tuple pattern cannot match values of the non-tuple type 'any Error'}}
     break
   }
   switch (x, 0) {
   case (.success, let y):
     break
+  // FIXME: We ought to improve the diagnostics here
   case (.failure(), let y):
     // expected-error@-1 {{tuple pattern cannot match values of the non-tuple type 'any Error'}}
     break
+  }
+
+  _ = {
+    switch x {
+    case .success():
+      // expected-error@-1 {{missing pattern #1 in enum associated value match}}
+      break
+    case .failure(let error):
+      break
+    }
   }
 }
 

--- a/test/Constraints/patterns.swift
+++ b/test/Constraints/patterns.swift
@@ -883,3 +883,15 @@ func testMissingPatternDiagnosticsInExpressionContext() {
     if case let .test(_, answer) = k {} // expected-error {{missing patterns #3, #4 in enum associated value match}} {{32-32=, <#Int#>, <#String#>}}
   }
 }
+
+// https://github.com/swiftlang/swift/issues/82520
+func testNestedMissingMembers() {
+  enum E {
+    case e
+  }
+
+  func foo(_ e: E) {
+    if case .a(.x) = e { // expected-error {{type 'E' has no member 'a'}}
+    }
+  }
+}

--- a/test/Constraints/protocols.swift
+++ b/test/Constraints/protocols.swift
@@ -526,7 +526,6 @@ do {
   }
 }
 
-// FIXME: These should all produce correct diagnostics
 do {
   protocol P {}
 
@@ -539,18 +538,21 @@ do {
   func f2() -> (any P)? {
     let x: Any? = nil
     return x
-    // expected-error@-1 {{failed to produce diagnostic for expression}}
+    // expected-error@-1 {{cannot convert return expression of type 'Any?' to return type '(any P)?'}}
+    // expected-note@-2 {{arguments to generic parameter 'Wrapped' ('Any' and 'any P') are expected to be equal}}
   }
 
   func f3() -> [any P] {
     let x: [Any] = []
     return x
-    // expected-error@-1 {{failed to produce diagnostic for expression}}
+    // expected-error@-1 {{cannot convert return expression of type '[Any]' to return type '[any P]'}}
+    // expected-note@-2 {{arguments to generic parameter 'Element' ('Any' and 'any P') are expected to be equal}}
   }
 
   func f4() -> [Int: any P] {
     let x: [Int: Any] = [:]
     return x
-    // expected-error@-1 {{failed to produce diagnostic for expression}}
+    // expected-error@-1 {{cannot convert return expression of type '[Int : Any]' to return type '[Int : any P]'}}
+    // expected-note@-2 {{arguments to generic parameter 'Value' ('Any' and 'any P') are expected to be equal}}
   }
 }

--- a/test/Constraints/result_builder_diags.swift
+++ b/test/Constraints/result_builder_diags.swift
@@ -1109,3 +1109,19 @@ func testInvalidTransforDoesntHideConversionFailure() {
     test("") // expected-error {{cannot convert value of type 'String' to expected argument type 'Int'}}
   }
 }
+
+func testInvalidPatternInResultBuilderBody() {
+  @resultBuilder struct Builder {
+    static func buildBlock() -> Int { 0 }
+  }
+
+  struct Test {
+    @Builder func invalidA() -> Int {
+      let a // expected-error {{type annotation missing in pattern}}
+    }
+
+    @Builder var invalidB: Int {
+      @Invalid var b // expected-error {{type annotation missing in pattern}}
+    }
+  }
+}

--- a/test/Sema/object_literals_osx.swift
+++ b/test/Sema/object_literals_osx.swift
@@ -31,7 +31,8 @@ let p1: Path = #fileLiteral(resourceName: "what.txt")
 let p2 = #fileLiteral(resourceName: "what.txt") // expected-error{{could not infer type of file reference literal}}
 // expected-note@-1{{import Foundation to use 'URL' as the default file reference literal type}}
 
-let text = #fileLiteral(resourceName: "TextFile.txt").relativeString! // expected-error{{failed to produce diagnostic for expression; please submit a bug report (https://swift.org/contributing/#reporting-bugs)}}
+let text = #fileLiteral(resourceName: "TextFile.txt").relativeString! // expected-error{{could not infer type of file reference literal}}
+// expected-note@-1{{import Foundation to use 'URL' as the default file reference literal type}}
 
 // rdar://problem/49861813
 #fileLiteral()

--- a/test/expr/expressions.swift
+++ b/test/expr/expressions.swift
@@ -795,8 +795,7 @@ func testNilCoalescePrecedence(cond: Bool, a: Int?, r: ClosedRange<Int>?) {
   // ?? should have higher precedence than logical operators like || and comparisons.
   if cond || (a ?? 42 > 0) {}  // Ok.
   if (cond || a) ?? 42 > 0 {}  // expected-error {{cannot be used as a boolean}} {{15-15=(}} {{16-16= != nil)}}
-  // expected-error@-1 {{binary operator '>' cannot be applied to operands of type 'Bool' and 'Int'}}  expected-note@-1 {{overloads for '>' exist with these partially matching parameter list}}
-  // expected-error@-2 {{binary operator '??' cannot be applied to operands of type 'Bool' and 'Int'}}
+  // expected-error@-1 {{cannot convert value of type 'Bool' to expected argument type 'Int'}}
   if (cond || a) ?? (42 > 0) {}  // expected-error {{cannot be used as a boolean}} {{15-15=(}} {{16-16= != nil)}}
 
   if cond || a ?? 42 > 0 {}    // Parses as the first one, not the others.

--- a/test/expr/leading_dot/static_var.swift
+++ b/test/expr/leading_dot/static_var.swift
@@ -65,3 +65,18 @@ struct ImplicitMembers {
 func implicit(_ i: inout ImplicitMembers) {
     if i =% .optional {}
 }
+
+do {
+  struct Property<R, T> {
+    static var `default`: Property<Loader, Int> { .init() }
+  }
+
+  struct Loader {
+    func load<T>(_: Property<Self, T>) -> T {
+    }
+  }
+
+  func test(loader: Loader) {
+    let _: Double = loader.load(.default) // expected-error {{cannot convert value of type 'Int' to specified type 'Double'}}
+  }
+}

--- a/test/type/opaque.swift
+++ b/test/type/opaque.swift
@@ -619,3 +619,19 @@ do {
   func f2() -> some P33333 {}
   // expected-error@-1 {{cannot find type 'P33333' in scope}}
 }
+
+do {
+  struct MyV : Proto {
+    func scaled() -> some Proto { MyV() }
+  }
+
+  struct Test<T: Proto> {
+    let v1: T?
+    let v2: T?
+  }
+
+  func test() {
+    Test(v1: MyV().scaled(), v2: MyV())
+    // expected-error@-1 {{conflicting arguments to generic parameter 'T' ('some Proto' (result type of 'scaled') vs. 'MyV')}}
+  }
+}

--- a/validation-test/Sema/SwiftUI/rdar84048788.swift
+++ b/validation-test/Sema/SwiftUI/rdar84048788.swift
@@ -1,0 +1,23 @@
+// RUN: %target-typecheck-verify-swift -target %target-cpu-apple-macosx10.15 -swift-version 5
+// REQUIRES: objc_interop
+// REQUIRES: OS=macosx
+
+import SwiftUI
+
+struct MyEditMode {
+}
+
+struct ContentView: View {
+  var editMode: MyEditMode?
+
+  var body: some View {
+    EmptyView()
+      .toolbar {
+        ToolbarItem(placement: .navigationBarTrailing) {
+          Button(editMode? ? "Done" : "Edit") { // expected-error {{optional type 'MyEditMode?' cannot be used as a boolean; test for '!= nil' instead}}
+            print("Pressed")
+          }
+        }
+      }
+  }
+}

